x/auth/ante/circuit.go

```diff
@@ -0,0 +1,22 @@
+package ante
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+)
+
+type circuitBreakerDecorator struct {
+	keeper CircuitBreakerKeeper
+}
+
+func (d *circuitBreakerDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bool, next sdk.AnteHandler) (newCtx sdk.Context, err error) {
+	if !d.keeper.IsCircuitOpen(ctx) {
```

Comment:
I don't think this is the expected behavior. In fact, see @aaronc's message [here](https://github.com/cosmos/cosmos-sdk/pull/14521/files#r1100270554).

What this should look like is:

```go
for _, msg := range tx.GetMsgs() {
  if !d.keeper.IsAllowed(ctx, msg) {
    return ctx, sdkerrors.Wrap(sdkerrors.MsgTripped, sdk.MessageTypeURL(msg))
  }
}

// transaction is valid and no messages are tripped
return next(ctx, tx, simulate)
```

What's happening here is that we need to check each message and ensure it's not tripped. IFF all messages are not tripped, do we consider this tx valid and OK.

Now, I also want to point out, due to x/authz, we might need to inspect messages directly. I.e. if it's a `MsgExec` we might need to dive into the exec messages.

cc @aaronc

Reply:
Yeah, this logic should probably be inside the MsgServiceRouter to handle cases like x/authz.

Reply:
You mean in `baseapp/msg_service_router.go`? So you don't think it should be an ante-handler at all?

Reply:
@aaronc can you elaborate? I'm a little confused as I thought that this had to go in the ante-handler?

Reply:
hmm this is the an issue with authz and how the router works, since they don't go through the ante handler. I believe the design here would one similar to consensus params, would have a IsAllowed expected interface in base app and then circuit breaker is passed into it in app.go.

Reply:
It's still not clear where this resides. Hence my question in https://github.com/cosmos/cosmos-sdk/pull/14521#discussion_r1101647776

Reply:
Would this not reside in `baseapp/circuit.go` ? Following the consensus params path

Reply:
the expected interface will live in baseapp, correct.